### ISSUE TITLE
Heap size OoM on Azure AKS Multi-Node Setup

### ISSUE THREAD

**Issue by tggreene** [xtdb/crux#3650]
We're seeing OOM issues on pods running the benchmark as found by Dan. Just spinning up the pods is all the repro needed for this.

## Edit - Dan M, copying over my original notes

Running a node with:
- 2Gb of Heap 
- 2Gb of MaxDirectMemorySize

We see errors within the multi node run for both of the following:
- `java.lang.OutOfMemoryError: Java heap space`
- `io.netty.util.internal.OutOfDirectMemoryError: failed to allocate 268435456 byte(s) of direct memory (used: 1933574151, max: 2097152000)`

We could/should investigate behaviour when bumping both of those values, but given that AWS seems to run fine with lower limits for both, suspect there could be a bit of profiling we may want to do against Azure to see what areas/parts are using the most memory (ie, is there anything in the object store which demands a lot more memory than say, S3)

**Comment by jarohen** [comment]
potentially blocked on #3629 as that's preventing the node from loading the initial data correctly

**Comment by tggreene** [comment]
Have spent some time looking into this and the piece of code that's implicated in all the out of heap exceptions we've observed is `xtdb.azure.object-store/get-blob`:

```
16:54:09.689 [ForkJoinPool.commonPool-worker-1] WARN  reactor.core.Exceptions - throwIfFatal detected a jvm fatal exception, which is thrown and logged below:
java.lang.OutOfMemoryError: Java heap space
	at java.base/java.util.Arrays.copyOf(Arrays.java:3541)
	at com.azure.core.implementation.ByteBufferCollector.toByteArray(ByteBufferCollector.java:95)
	at com.azure.core.util.FluxUtil$$Lambda/0x0000000101bd6700.apply(Unknown Source)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onNext(FluxMapFuseable.java:113)
	at reactor.core.publisher.Operators$MonoSubscriber.complete(Operators.java:1839)
	at reactor.core.publisher.MonoCollect$CollectSubscriber.onComplete(MonoCollect.java:160)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onComplete(FluxMapFuseable.java:152)
	at reactor.core.publisher.FluxIterable$IterableSubscription.fastPath(FluxIterable.java:424)
	at reactor.core.publisher.FluxIterable$IterableSubscription.request(FluxIterable.java:291)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.request(FluxMapFuseable.java:171)
	at reactor.core.publisher.MonoCollect$CollectSubscriber.onSubscribe(MonoCollect.java:104)
	at reactor.core.publisher.FluxMapFuseable$MapFuseableSubscriber.onSubscribe(FluxMapFuseable.java:96)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:201)
	at reactor.core.publisher.FluxIterable.subscribe(FluxIterable.java:83)
	at reactor.core.publisher.Mono.subscribe(Mono.java:4491)
	at reactor.core.publisher.NextProcessor.connect(NextProcessor.java:500)
	at reactor.core.publisher.NextProcessor.block(NextProcessor.java:112)
	at reactor.core.publisher.MonoProcessor.block(MonoProcessor.java:116)
	at com.azure.core.implementation.util.FluxByteBufferContent.getBytes(FluxByteBufferContent.java:206)
	at com.azure.core.implementation.util.FluxByteBufferContent.lambda$toBytes$0(FluxByteBufferContent.java:94)
	at com.azure.core.implementation.util.FluxByteBufferContent$$Lambda/0x0000000101bd5470.apply(Unknown Source)
	at java.base/java.util.concurrent.atomic.AtomicReferenceFieldUpdater.updateAndGet(AtomicReferenceFieldUpdater.java:240)
	at com.azure.core.implementation.util.FluxByteBufferContent.toBytes(FluxByteBufferContent.java:94)
	at com.azure.core.implementation.util.FluxByteBufferContent.toByteBuffer(FluxByteBufferContent.java:109)
	at com.azure.core.util.BinaryData.toByteBuffer(BinaryData.java:1411)
	at xtdb.azure.object_store$get_blob.invokeStatic(object_store.clj:23)
	at xtdb.azure.object_store$get_blob.invoke(object_store.clj:19)
	at xtdb.azure.object_store.AzureBlobObjectStore$reify__31316.get(object_store.clj:127)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1768)
	at java.base/java.util.concurrent.CompletableFuture$AsyncSupply.exec(CompletableFuture.java:1760)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:387)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1312)
```

The possibilities are:
 - We're not using the SDK correctly (although I believe we are) and memory is not being freed
 - The SDK has some issue causing memory to not be freed
 - We're allocating too many objects (possibly too large too) at once

I've set off another 24H run with heapdump directly correctly set so we might get more evidence of this being the source of the issue.

Otherwise I'm looking at trying to replicate this locally.
Alternatively, it may just be the case that we need a bigger heap for azure.

**Comment by tggreene** [comment]
These were the logs of the pods that failed:

[xtdb-azure-bench-heap-space-1.log](https://github.com/user-attachments/files/16944347/xtdb-azure-bench-heap-space-1.log)
[xtdb-azure-bench-heap-space-2.log](https://github.com/user-attachments/files/16944348/xtdb-azure-bench-heap-space-2.log)
[xtdb-azure-bench-unsafe-memory-access.log](https://github.com/user-attachments/files/16944349/xtdb-azure-bench-unsafe-memory-access.log)

**Comment by danmason** [comment]
Relating to the heap size OoM:

Just pushed a change to `main` to directly download the arrow files to a local `out-path`, rather than first downloading them as a byte buffer and writing the byte buffer to a file. The current theory being that the increased memory pressure may result from parallel calls of `getBlob` loading multiple arrow files into memory simultaneously, causing us to go above our heap size limits. We shall push this change to the azure container registry and run a multiple-node benchmark against it, observing the behaviour.

**Comment by danmason** [comment]
We've yet to see another heap size OoM since pushing this (though we did encounter slight other issues with the code, see the PR #3692 ), when running the multi nodes for an hour. With the cleanup/change here to ensure that the threads don't fail/die out now pushed on the image, I've kicked off a 6 hour run to observe the behaviour on the multiple nodes to see if we encounter any further Heap Size OoMs (not counting the direct memory ones - these seem to be separate and worth another card). 

If the 6 hour run goes okay & we're happy that we haven't seen any on Azure for a while, shall consider this closed.
